experimental/nomad_resource_plugin/licensedevice/plugin.go

```diff
@@ -40,3 +90,83 @@ func (p *Plugin) Reserve(deviceIDs []string) (*device.ContainerReservation, erro
 func (p *Plugin) Stats(ctx context.Context, interval time.Duration) (<-chan *device.StatsResponse, error) {
 	return nil, fmt.Errorf("Stats not implemented")
 }
+
+func (p *Plugin) fingerprintLoop(ctx context.Context, notifyChan <-chan struct{}, resChan chan<- *device.FingerprintResponse) {
+	for {
```

Comment:
As far as I can tell, this is constantly polling.
Feels like with any notifier we supply we'll want some purposeful delay between polls as having potentially 10s or 100s of machines continually polling the same notifier could have scale issues.

If you agree with the above, then my assumption would be that we could put some rate-limiting in the notifier code. However we set a 5-second timeout in this loop when we call the notifier. So if the notifier rate-limited, this timeout wouldn't place nicely with it.

All of the above to ask:

What's your plan for rate-limiting and timeout handling?
If you agree with adding it, are you thinking about adding it here or in the notifier?

Reply:
Each nomad client needs to run a copy of this application, so each will have a connection to the DB for listening for updates, and each will be watching its own `/proc/mounts` file. In that way, multiple machines aren't using a single notifier object (because each machine is running its own instance of this plugin).

The notifier interface is supposed to discourage polling; instead, it offers a channel that signals that a change has happened, and the top-level code should go re-read. Under the hood, the notifier impl could be polling, or it could have some push-based mechanism.

For the DB, postgres has `LISTEN`/`NOTIFY` that we can use to implement this without doing e.g. a SELECT loop: https://www.postgresql.org/docs/16/sql-notify.html

For files, specifically `/proc/mounts`, it's less clear what to use, though it sounds like `inotify` doesn't work for proc files; the alternative is a syscall called (confusingly) `poll`: https://linux.die.net/man/2/poll (basically, register to receive events on a set of file descriptors)

---

The interface is currently a function that returns a channel of empty struct (notification only, no data) + a function to "get current" - I'm not sure why it can't be a channel of current data (so each channel receive implies an event), and I might change it to that in a future PR.

Reply:
Okay so in this case, the fingerprint loop is waiting for a notifier to trigger, that there was a change (notifyChan)
Then this code goes to that notifier, gets the current state, does some processing, and signals the current state to the resChan (Nomad)